### PR TITLE
Simple powered up led

### PR DESCRIPTION
**Pull request by topas-rec**
In my opinion a good addition to find out if usb power connection can be safely disconnected between Raspberry Pi and the Power bank using an external switch.
Led is turned on, when Pi is running, when pushing power button is blinks two times to indicate shutdown initiated. Then is stays on until Pi loses power. Now one can disconnect power safely.

Led uses free GPIO 3.
Can be ignored if led is not used.

Renamed GPIO buttons file because it now also controls led.

I made video of the functionality, but need to upload it somewhere.

### CODE REVIEW

**Comment by topas-rec** [comment]
I think this will not destroy the "no soldering" solution, because one can easily ignore the led like the gpio buttons.

**Comment by topas-rec** [comment]
I have not checked the update conditions in detail. I fyou want to merge we should concider the renamed file (gio-buttons.py to gpio.py) in the install script(s).

**Comment by topas-rec** [comment]
This is a small video of the functionality.

[VID_20181215_105739.m4v.zip](https://github.com/MiczFlor/RPi-Jukebox-RFID/files/2682559/VID_20181215_105739.m4v.zip)

File: scripts/installscripts/stretch-install-default.sh [hunk 1]
```diff
@@ -482,30 +482,30 @@ echo "### Deleting older versions of service daemons. This might throw errors, i
 sudo systemctl disable idle-watchdog
 sudo systemctl disable rfid-reader
 sudo systemctl disable startup-sound
-sudo systemctl disable gpio-buttons
```
Comment (topas-rec) [inline scripts/installscripts/stretch-install-default.sh]: This should stay for a while I guess right?

File: scripts/installscripts/stretch-install-default.sh [hunk 2]
```diff
@@ -482,30 +482,30 @@ echo "### Deleting older versions of service daemons. This might throw errors, i
...
+sudo systemctl disable gpio
 sudo rm /etc/systemd/system/rfid-reader.service 
 sudo rm /etc/systemd/system/startup-sound.service
-sudo rm /etc/systemd/system/gpio-buttons.service
```
Comment (topas-rec) [inline scripts/installscripts/stretch-install-default.sh]: This should stay for a while I guess right?

**Comment by martinclausen2** [comment]
Hi @topas-rec, GPIO 3 is not free, it is currently GPIO to power up and down and can not be chnaged to other pins (wake-up is special pin requirement).
Please find a suitable pin by considering the [new pinout](https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/Hardware-Pinout-Overview)

**Comment by MiczFlor** [comment]
Hi @topas-rec @martinclausen2 
thanks for the contribution and thanks for pointing out the conflict. Is there a way to modify the code to follow the GPIO "best practice" ideas?
* https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/Hardware-Pinout-Overview
* https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/Using-GPIO-hardware-buttons

**Comment by martinclausen2** [comment]
Hi @topas-rec , a suggestion GPIO 7 is free.

**Comment by topas-rec** [comment]
Cab you confirm that and then tell me if this gpio is free?
This gpio.py script works on my phoniebox together with the gpio buttons exactly implemented according to the wiki suggestion: https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/Using-GPIO-hardware-buttons

That's why I didn't believe there is a conflict between the led and the shutdown button, because I use them together.

**Comment by topas-rec** [comment]
OK found out that the gpio 2 is a conflict for the i2c protocol for the hifiberry. Correct?
So changing to gpio 7 is still necessary right?

**Comment by MiczFlor** [comment]
Hi @topas-rec 
this is still pending for two reasons. The first one I can not solve: I have not GPIO installed at the moment and can not test nor verify which pin should be used.
Secondly, you are right, the name of the script *should* change, because it's not only about buttons anymore. However, I do not want to deal with the disappointed parents after a `git pull` - which is why I would like to leave the filename as it is now.
All the best, micz

**Comment by topas-rec** [comment]
Hey @MiczFlor,

First:
> I have not GPIO installed at the moment and can not test nor verify which pin should be used.
What can I do now? As far as I know current gpio 2 is not ideal. Would it help to change to gpio 7?

Second:
>I do not want to deal with the disappointed parents after a git pull - which is why I would like to leave the filename as it is now.
Understand. I think a rename wouldn't break other installations, but no problem. I'll change back to original file name.

I'll change this soon.

**Comment by martinclausen2** [comment]
Hi @topas-rec Hi @MiczFlor GPIO 7 is fine, just use that. Regards, Martin

**Comment by topas-rec** [comment]
Functional tests not done.
Please wait for test result before merging.

File: misc/sampleconfigs/gpio-buttons.py.sample [hunk 3]
```diff
@@ -20,6 +20,12 @@ from subprocess import check_call
 # If anybody has ideas or tests or experience regarding this solution, please create pull requests or contact me.
 
 def def_shutdown():
+    # Power led blinks two times - Tell listener we're going down.
+    pled.blink(on_time=0.3, off_time=0.3, n=2, background=False)
+    # Power led stays on until shutdown completes (when blinking it might be
```
Comment (topas-rec) [inline misc/sampleconfigs/gpio-buttons.py.sample]: Make this easier to understand

File: misc/sampleconfigs/gpio-buttons.py.sample [hunk 4]
```diff
@@ -60,4 +67,7 @@ next.when_pressed = def_next
 prev.when_pressed = def_prev
 halt.when_pressed = def_halt
 
+# Power led running all the time
```
Comment (topas-rec) [inline misc/sampleconfigs/gpio-buttons.py.sample]: .... After startup and during normal operation

**Comment by topas-rec** [comment]
Ready to merge in my eyes.
Tested using an install on an existing and installed box without keeping the set up configs. (Fresh install selected)

Please merge and if possible add LED Pin description to GPIO-Buttons in the Wiki. I think I don't have acces to it.

**Comment by topas-rec** [comment]
Led pin neds to be added in the GPIO conflicts table, too.

**Comment by martinclausen2** [comment]
@topas-rec you have the rights, please edit the wiki yourself

**Comment by martinclausen2** [comment]
@topas-rec yes you can, please add your pin [here](https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/Hardware-Pinout-Overview#flexible-pinout)

**Comment by martinclausen2** [comment]
@topas-rec  After reviewing your code I understood that you implemented a similar feature as I did for the encoder buttons example. Therefore I would like to ask you to move to GPIO 12 which is already assigned to "Status LED".
An additional comment: your solution will not work well for Pi Zero users because gpiozero is not light wight enough to be available early enough in the system startup. [Details](https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/397)
Could you refactor to use plain RPi.GPIO instead? maybe we should also have a common base class to avoid code duplication.

**Comment by topas-rec** [comment]
Moving to GPIO shouldn't be the problem.

Using RPi.GPIO seems also pretty easy according to https://sourceforge.net/p/raspberry-gpio-python/wiki/BasicUsage/

Let's see if I can make it work using that python package.

**Comment by topas-rec** [comment]
I have no idea what you mean with a common base class. (I know what a base class is but we only have python scripts and no classes as far as I know)

**Comment by martinclausen2** [comment]
My head is stuck in C# I admit. I would just propose to adhere to the do not repeat your self (nor others ;-) )

Please see (my implementation)[https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/master/misc/sampleconfigs/gpio-buttons.py.rotaryencoder.sample] - I am always open for feedback. Maybe we can get also rid of the repeated parts of the path (line 19 / 22 in my code, many lines more in the code you are working on)

**Comment by burberius** [comment]
I tried that MR in my installation (just changed the LED GPIO to 12) and it works like expected!

**Comment by Joggel84** [comment]
I tried the new added lines from the 07.Feb today to check if it works well as Power ON Led.
And what should i Say it works very well!!

**Comment by topas-rec** [comment]
Thaanks for testing and for letting us know.
Have you tested the buttons, too? Do the still work?

**Comment by topas-rec** [comment]
I haven't tested this code after the changes from 7 th. February.

**Comment by Joggel84** [comment]
I'm sorry i did'nt test the buttons, because i don't need them. Maybe i use the shutdown button later on.
Aktual i try to realise some "light show" with some LED's at GPIO

**Comment by Joggel84** [comment]
But what i see at the moment the documantaion on 
https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/Hardware-Pinout-Overview
doesn't match to the   gpio-buttons.py

**Comment by martinclausen2** [comment]
Could you be please more explicit? Note, that there are two different pinouts suggested old one with collisions and new one without but with breaking changes 
You have to die one way or the other ;-) 

Am 17. März 2019 13:38:02 MEZ schrieb Joggel84 <notifications@github.com>:

**Comment by topas-rec** [comment]
I tried to leave the pins as they were before the changes from 2c4fa26 
Before that we talked about incompatible pins as you can see above in this discussion.

@martinclausen2 were you able to test this already?

**Comment by martinclausen2** [comment]
@Joggel84 The code follows the classic pinout and is aligned with the documentation so far, the status LED is not defined there. In this case i generally would recommend that the code should follow the documentation and not the other way around as we try to get everything in sync. IMHO In a perfect world we would all switch to the new flexible pinout.
@topas-rec I cannot test the code, because my hardware follows the new layout and I do not have buttons except for power and play/pause, the other functions are accessible via encoders
I see myself here as the one trying to moderate pinouts and take care of the encoder stuff.

**Comment by winterrace** [comment]
Why not simply forward the Pi's onboard LEDs with fibre optic cables?

light

Works great in my phoniebox :)

**Comment by martinclausen2** [comment]
@winterrace I had that idea as well, but how do you mount the light guide onto the leds on the RPi? Glue + hole in the case? After all I found it easier to redirect the LEDs onto a GPIO and add a LED to it.

**Comment by winterrace** [comment]
I drilled 2 holes into the upper shell of the pi case right above the LEDs. Wide enough that the fibre cable can easily be inserted. 
Then i temporarily covered these holes with tape and placed the lid upside down. I filled the entire edge with hot glue. After it cooled down I drilled 2 holes through the existing case holes through the glue layer (which is ca. 1 cm thick) towards the LEDs. I used a slightly smaller size of the drill bit when drilling through the glue layer so that the fibre cable will fit tight in that tunnel later. 

This way it's not necessary to apply any glue etc. to the pi itself. Just push the fibre cable through the glue tunnel and the last 5 mm free space until it attaches itself right on the LED surface.

And the fibres can easily be removed with the lid and be attached again with same precision.

**Comment by dvalob** [comment]
Excellent idea!!

**Comment by MiczFlor** [comment]
Hi @winterrace 
very sweet idea! I added it to the wiki:
https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/Forward-Pi-onboard-LEDs-with-fibre-optics

**Comment by topas-rec** [comment]
Tests finished. Works (on my machine)

**Comment by topas-rec** [comment]
If one merges: Stash all my commits into one!

**Comment by topas-rec** [comment]
Somehow after some time of use there is "No space left on device". I have to check if that has to do with this PR.

**Comment by topas-rec** [comment]
I left over a print statement which bloated my logs. Is not committed, was only on my local device.

Holding one of my volume buttons sometimes doesn't work when pushing the other one once right before holding the first one. Can someone test that?

**Comment by topas-rec** [comment]
Have done a fresh install for fun now. I'll give this piece of software to my 4 year old tester for one or two weeks now. Then we can decide to merge or modify.
Again please test if possible.

`cd; rm stretch-install-*; wget https://raw.githubusercontent.com/topas-rec/RPi-Jukebox-RFID/master/scripts/installscripts/stretch-install-default.sh; chmod +x stretch-install-default.sh; ./stretch-install-default.sh`

**Comment by topas-rec** [comment]
Still stable no issues. Tests done by me and my son.

**Comment by kabakakao** [comment]
I really like this addition and I would like to see this PR merged

**Comment by topas-rec** [comment]
@kabakakao, if you've testes this PR, can you provide us your experience? Are there issues with this PR?

**Comment by kabakakao** [comment]
I will test it the next days. I will comment this thread

**Comment by kabakakao** [comment]
I've tested the new gpio-file for my power LED. It works like a charm.
Thanks for that. Unfortunately i have no other gpio buttons to test them. But as I reviewed the changes, it should work. The other buttons are just on another gpio. An they seem to be unchanged.

File: misc/sampleconfigs/gpio-buttons.py.sample [hunk 5]
```diff
@@ -19,59 +25,115 @@ from subprocess import check_call
...
-
-def def_recordplaylatest():
+def recordplaylatest_action(channel):
     check_call("./scripts/playout_controls.sh -c=recordplaylatest", shell=True)
 
-shut = Button(3, hold_time=2)
-vol0 = Button(13,pull_up=True)
-volU = Button(16,pull_up=True,hold_time=0.3,hold_repeat=True)
-volD = Button(19,pull_up=True,hold_time=0.3,hold_repeat=True)
-next = Button(26,pull_up=True)
```
Comment (kabakakao) [inline misc/sampleconfigs/gpio-buttons.py.sample]: change it to (#694 ):
`fwfd = Button(26,pull_up=True)`

File: misc/sampleconfigs/gpio-buttons.py.sample [hunk 6]
```diff
@@ -19,59 +25,115 @@ from subprocess import check_call
...
+volumeHoldTime = 0.3 # Seconds
+shutdownHoldTime = 2 # Seconds
+PledBlinkTime = 0.3  # Seconds
+
+# Set up GPIO pins and the power led
+GPIO.setup(shut, GPIO.IN, pull_up_down=GPIO.PUD_UP)
+GPIO.setup(vol0, GPIO.IN, pull_up_down=GPIO.PUD_UP)
+GPIO.setup(volU, GPIO.IN, pull_up_down=GPIO.PUD_UP)
+GPIO.setup(volD, GPIO.IN, pull_up_down=GPIO.PUD_UP)
+GPIO.setup(next, GPIO.IN, pull_up_down=GPIO.PUD_UP)
```
Comment (kabakakao) [inline misc/sampleconfigs/gpio-buttons.py.sample]: change it to (#694 ):
` GPIO.setup(fwfd, GPIO.IN, pull_up_down=GPIO.PUD_UP)`

File: misc/sampleconfigs/gpio-buttons.py.sample [hunk 7]
```diff
@@ -19,59 +25,115 @@ from subprocess import check_call
...
+    # Detect holding of button
+    while checkGpioStaysInState(volumeHoldTime, volD, GPIO.LOW):
+        volD_action(channel)
+
+# Define the used pins of the raspberry board
+shut = 3
+vol0 = 13
+volU = 16
+volD = 19
+next = 26
```
Comment (kabakakao) [inline misc/sampleconfigs/gpio-buttons.py.sample]: change it to (#694 ):
`fwrd = Button(26,pull_up=True)`

File: misc/sampleconfigs/gpio-buttons.py.sample [hunk 8]
```diff
@@ -19,59 +25,115 @@ from subprocess import check_call
...
+GPIO.setup(prev, GPIO.IN, pull_up_down=GPIO.PUD_UP)
+GPIO.setup(halt, GPIO.IN, pull_up_down=GPIO.PUD_UP)
+GPIO.setup(pled, GPIO.OUT)
+
+# Set standard events for the buttons. Callback functions define the actions of the events (THey are defined above)
+GPIO.add_event_detect(shut, GPIO.FALLING, callback=shutdown_handler, bouncetime=bouncetime)
+GPIO.add_event_detect(vol0, GPIO.FALLING, callback=vol0_action, bouncetime=bouncetime)
+GPIO.add_event_detect(volU, GPIO.FALLING, callback=volU_handler, bouncetime=bouncetime)
+GPIO.add_event_detect(volD, GPIO.FALLING, callback=volD_handler, bouncetime=bouncetime)
+GPIO.add_event_detect(next, GPIO.FALLING, callback=next_action, bouncetime=bouncetime)
```
Comment (ctietze) [inline misc/sampleconfigs/gpio-buttons.py.sample]: change _next_ see above comments
`GPIO.add_event_detect(fwrd, GPIO.FALLING, callback=next_action, bouncetime=bouncetime)`

**Comment by patrickweigelt** [comment]
It also works to connect an LED to an usb-port and it will light up when booting and turn off when shut down without any code. See here: https://twitter.com/Weigelt_P/status/1217192571346259969

**Comment by topas-rec** [comment]
I changed the base branch to be able to integrate the changes mentioned by @kabakakao regarding the python
`next`
keyword.

I will adopt my branch to the develop branch soon.

**Comment by s-martin** [comment]
If I remember correctly renaming of `next` was already done in develop branch.

**Comment by topas-rec** [comment]
> If I remember correctly renaming of next was already done in develop branch.
Jepp. Thats why I changed to the develop branch. Getting this fix into this PR can be done in one shot instead of doing it later when merging develop in master.

**Comment by topas-rec** [comment]
This PR was targeted to master. In master the `next` changes are not merged.

**Comment by topas-rec** [comment]
- LED changed to GPIO 12
- Renaming `next` variables done

I'll test this soon. If someone can test it that would help.
I saw the rfid switch off function and adopted it to work with `RPi.GPIO`. @tobias-snapit might want to test that. (It is commented out - as it was before)

I think this can be merged already since the GPIO and LED code was already tested before.

**Comment by topas-rec** [comment]
Tested with current buster install script (develop branch) and with GPIO 12.

Works right after install and works as expected.

Please merge.

**Comment by veloxidSchweiz** [comment]
Hi @topas-rec ,
I am wondering why you removed the gpiozero module?
Isn't it possible to implement it with gpiozero?
actually another nice way to do this would be to use a 
dtoverlay:
`dtoverlay=gpio-poweroff,gpiopin=13`
would power on gpiopin 13 (if i remember correctly BCM) as soon as power is down and the rpi can be put off. 
It can als be used in inverted mode by using `,active_low="y"`....

**Comment by martinclausen2** [comment]
Hi @veloxidSchweiz a year oago I removed carefully gpio.zero for performance reasons:
https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/387#issuecomment-454953113
As you noticed here https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/763 it is advisable to use less different libs for the same task. So we agreed on dropping gpiozero at the time to my understanding. If you use a RPi 4 you will likely not notice, but for a Pi Zero it is a different story.

**Comment by veloxidSchweiz** [comment]
Hi @martinclausen2 ,
Thanks for the explanation, I wasn't aware of that.
I am usually not using the Pi Zero nor a RPi4 but RPI 3 therefore i probably haven't recognized it. 
At the end i think it is the best at least to agree on one interface. 
After reading the discussions here, I am asking me if the idea of the common base class goes somehow in the same direction I am suggesting in #763 ...

**Comment by topas-rec** [comment]
Can someone of the maintainers please explain to me why this is not going to be merged or rejected? I see a lot of discussion going on around config files and reorganizing the project. I think that is all great. Does it has to do with this PR? Is that the reason that this PR is not merged? Are there other reasons?

This PR is open for a year now and I am afraid someday I find a whole different project and my changes do not fit anymore and I can rewrite this (small) change. I don't have so much time so I am not sure to be able to keep up with the projects speed.

**Comment by MiczFlor** [comment]
Hi @topas-rec 
you are absolutely right. Did you rebase your work to the latest version of the script? I will merge it immediately. I am not super up to date with the status of your contribution. Apologies!
All the best, micz

**Comment by topas-rec** [comment]
I did not rebase but I merged your development branch into this branch just now.
So yes, in my branch are the latest changes of the script.

**Comment by topas-rec** [comment]
@MiczFlor everything is up to date - please merge.

**Comment by narf0815** [comment]
is there a way to simply implement this in stable?
as mentioned [here](https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/900#issue-599793406) i have some trouble with developer...

(am i supposed to post here?)

**Comment by topas-rec** [comment]
It has to be merged by a maintainer.

You can copy the file gpio-buttons.py from develop to your box manually until it is merged.

**Comment by narf0815** [comment]
i tried my working develop file to my stable version but LED does still not work on stable.

**Comment by topas-rec** [comment]
Led connected as described here?
https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/Using-GPIO-hardware-status-led

**Comment by narf0815** [comment]
yes. and it's working on my "develop-sdcard"

**Comment by topas-rec** [comment]
So just not working when copying the file gpio-buttons.py into the stable (master) branch?

I don't have an idea why. For me this file was always like an atomic thing. Works when present.

Can you put a debug printf in that file to see what that code does and why the led doesn't light up?

You can see these printf by looking in the /var/logsyslog file.
You can use the tail command for automatically refreshing through SSH session.

**Comment by narf0815** [comment]
> So just not working when copying the file gpio-buttons.py into the stable (master) branch?

exactly
> Can you put a debug printf in that file to see what that code does and why the led doesn't light up?

huh. if you tell me where to put what i shure could.

**Comment by topas-rec** [comment]
General printing of messages in python:
https://www.w3schools.com/python/ref_func_print.asp

**Comment by topas-rec** [comment]
@MiczFlor let us know if this should better be discussed in an issue rather than in a merged PR.

**Comment by narf0815** [comment]
i'm afraid this is (far) above my "potential".
hope this comes to stable soon so i/my kids can have volume rotary encoder and LED.

thanks for your effort however!

**Comment by s-martin** [comment]
We're working on releasing 2.0 "soon" although I can't give you a date. Please see #890 for coordination of the release.

**Pull request merged**